libcxx/include/__flat_map/flat_map.h

```diff
@@ -0,0 +1,1367 @@
+        // because we have two containers. The only possible solution is to clear them out
+        clear();
+        throw __flat_map_restore_error(
+            std::string("flat_map::emplace: Emplacement on the underlying container has failed and has side effect. "
+                        "Unable to restore flat_map to previous state. Clear out the containers to make the two "
+                        "containers consistent. Reason: ") +
+            __ex.what());
+      } catch (...) {
+        // The container might be in some unknown state and we can't get flat_map into consistent state
+        // because we have two containers. The only possible solution is to clear them out
+        clear();
+        throw __flat_map_restore_error(
+            "flat_map::emplace: Emplacement on the underlying container has failed and has side effect. "
+            "Unable to restore flat_map to previous state. Clear out the containers to make the two "
+            "containers consistent.");
+      }
+#  endif // _LIBCPP_HAS_NO_EXCEPTIONS
+    }
+  }
+
+  template <class _Container, class _Iter>
+  _LIBCPP_HIDE_FROM_ABI auto __safe_erase(_Container& __container, _Iter&& __iter) {
+    // [container.reqmts] No erase(), clear(), pop_back() or pop_front() function throws an exception,
+    // except that there are exceptional cases
+
+    // http://eel.is/c++draft/deque.modifiers#5
+    // http://eel.is/c++draft/vector.modifiers#4
+
+#  ifndef _LIBCPP_HAS_NO_EXCEPTIONS
+    try {
+#  endif // _LIBCPP_HAS_NO_EXCEPTIONS
+      return __container.erase(__iter);
+#  ifndef _LIBCPP_HAS_NO_EXCEPTIONS
+    } catch (const exception& __ex) {
+      // The container might be in some unknown state and we can't get flat_map into consistent state
+      // because we have two containers. The only possible solution is to clear them out
+      clear();
+      throw __flat_map_restore_error(
+          std::string("flat_map: Erasing on the underlying container has failed. "
+                      "Unable to restore flat_map to previous state. Clear out the containers to make the two "
+                      "containers consistent. Reason: ") +
+          __ex.what());
+    } catch (...) {
+      // The container might be in some unknown state and we can't get flat_map into consistent state
+      // because we have two containers. The only possible solution is to clear them out
+      clear();
+      throw __flat_map_restore_error(
+          "flat_map: Erasing on the underlying container has failed. "
+          "Unable to restore flat_map to previous state. Clear out the containers to make the two "
+          "containers consistent.");
+    }
+#  endif // _LIBCPP_HAS_NO_EXCEPTIONS
+  }
+
+  template <class _IterK, class _IterM, class _KeyArg, class... _MArgs>
+  _LIBCPP_HIDE_FROM_ABI iterator
+  __try_emplace_impl(_IterK&& __it_key, _IterM&& __it_mapped, _KeyArg&& __key, _MArgs&&... __mapped_args) {
```

Comment:
From our live review just now:

```c++
  template <class _IterK, class _IterM, class _KeyArg, class... _MArgs>
  _LIBCPP_HIDE_FROM_ABI iterator
  __try_emplace_impl(_IterK&& __it_key, _IterM&& __it_mapped, _KeyArg&& __key, _MArgs&&... __mapped_args) {
    auto __on_key_failed = std::__make_exception_guard([&]() noexcept {
      if constexpr (__emplacement_has_strong_exception_safety_guarantee<_KeyContainer>()) {
        // Nothing to roll back!
      } else {
        // we need to clear both because we don't know the state of our keys anymore
        clear() /* noexcept */;
      }
    });
    auto __key_it = __containers_.keys.emplace(__it_key, std::forward<_KeyArg>(__key));
    __on_key_failed.__complete();

    auto __on_value_failed = std::__make_exception_guard([&]() noexcept {
      if constexpr (!__emplacement_has_strong_exception_safety_guarantee<_ValueContainer>()) {
        // we need to clear both because we don't know the state of our values anymore
        clear() /* noexcept */;
      } else {
        // In this case, we know the values are just like before we attempted emplacement,
        // and we also know that the keys have been emplaced successfully. Just roll back the keys.
        try {
          __containers_.keys.erase(__key_it);
        } catch (...) {
          // Now things are funky for real. We're failing to rollback the keys.
          // Just give up and clear the whole thing.
          //
          // Also, swallow the exception that happened during the rollback and let the
          // original value-emplacement exception propagate normally.
          clear() /* noexcept */;
        }
      }
    });
    auto __mapped_it = __containers_.values.emplace(__it_mapped, std::forward<_MArgs>(__mapped_args)...);
    __on_value_failed.__complete();

    return iterator(std::move(__key_it), std::move(__mapped_it));
  }
```